File: courses/LEETCODE In Python 50 Algorithms Coding Interview Questions/10_math_question_count_primes_(easy)/better.py
"""Count Primes"""

import math


class Solution:
    def countPrimes(self, n):
        if n < 2:
            return 0
        isPrime = [True] * n
        isPrime[0] = isPrime[1] = False

        for i in range(2, int(math.ceil(math.sqrt(n)))):
            if isPrime[i]:
                for multiples_of_i in range(i * 2, n, i):
                    isPrime[multiples_of_i] = False

        primeCount = 0
        for i in range(n):
            if isPrime[i]:
                primeCount += 1

        return primeCount


if __name__ == "__main__":
    print(Solution().countPrimes(1))
    print(Solution().countPrimes(5))
    print(Solution().countPrimes(10))
    print(Solution().countPrimes(17))

File: courses/LEETCODE In Python 50 Algorithms Coding Interview Questions/10_math_question_count_primes_(easy)/better.js

/*  Given a number n, return an array of prime numbers whose
    size is equal to the number n. 

    Example: 
    0 -> [], 
    1 -> [2], 
    3 -> [2, 3, 5] 
*/

function getPrimes(num) {
    if (num === 0) {
        return []
    }
    if (num === 1) {
        return [2]
    }

    const length = num ** 2
    const isPrime = new Array(length).fill(true)
    isPrime[0] = false
    isPrime[1] = false
    let count = 0
    for (let i = 2; i < Math.ceil(Math.sqrt(length)); i++) {
        if (isPrime[i]) {
            for (let j = i * 2; j < length; j+= i) {
                isPrime[j] = false
            }
            count++
        }
        if (count === num) break
    }

    count = 0
    let primes = []
    for (let i = 0; i < length; i++) {
        if (isPrime[i]) {
            primes.push(i)
            count++
        }
        if (count === num) break
    }
    
    return primes
}
